services/codecov/codecov.service.js

```diff
@@ -1,108 +1,139 @@
 'use strict'
 
-const queryString = require('query-string')
-const LegacyService = require('../legacy-service')
-const { makeBadgeData: getBadgeData } = require('../../lib/badge-data')
-const {
-  coveragePercentage: coveragePercentageColor,
-} = require('../color-formatters')
+const Joi = require('joi')
+const { coveragePercentage } = require('../color-formatters')
+const { BaseJsonService } = require('..')
 
-// This legacy service should be rewritten to use e.g. BaseJsonService.
-//
-// Tips for rewriting:
-// https://github.com/badges/shields/blob/master/doc/rewriting-services.md
-//
-// Do not base new services on this code.
-module.exports = class Codecov extends LegacyService {
+// https://docs.codecov.io/reference#totals
+// A new repository that's been added but never had any coverage reports
+// uploaded will not have a `commit` object in the response and some times
+// the `totals` object can also be missing for the latest commit.
+// Keeping these as optional in the schema and handling it in the transform
+// function to maintain consistent badge behavior with the legacy service implementation.
+const schema = Joi.object({
+  commit: Joi.object({
+    totals: Joi.object({
+      c: Joi.number().required(),
+    }),
+  }),
+}).required()
+
+const queryParamSchema = Joi.object({
+  token: Joi.string(),
+}).required()
+
+const documentation = `
+  <p>
+    You may specify a Codecov token to get coverage for a private repository.
+  </p>
+  <p>
+    See the <a href="https://docs.codecov.io/reference#authorization">Codecov Docs</a>
+    for more information about creating a token.
+  </p>
+`
+
+module.exports = class Codecov extends BaseJsonService {
   static get category() {
     return 'coverage'
   }
 
+  static get defaultBadgeData() {
+    return { label: 'coverage' }
+  }
+
+  static render({ coverage }) {
+    if (coverage === 'unknown') {
+      return {
+        message: coverage,
+        color: 'lightgrey',
+      }
+    }
+    return {
+      message: `${coverage.toFixed(0)}%`,
+      color: coveragePercentage(coverage),
+    }
+  }
+
   static get route() {
     return {
       base: 'codecov/c',
-      pattern: '',
+      // https://docs.codecov.io/docs#section-common-questions
+      // Github, BitBucket, and GitLab are the only supported options (long or short form)
+      pattern:
+        ':vcsName(github|gh|bitbucket|bb|gl|gitlab)/:user/:repo/:branch*',
```

Comment:
I was torn on whether to include both the long and shorthand value for the `vcsName`. The legacy service regex would let either through, so I felt that we'd need to support both long/short forms to avoid breaking any existing users. 

If we prefer one form vs. the other, then I could probably set up a few redirectors to maintain backwards compatibility with the other form. However, I elected to stick with the broader route param support here since there's already a redirector transforming the legacy token route parts to a query param.

Let me know what folks think

Reply:
I'm inclined to standardize on the longer forms at some point, though I don't think it's important to create more redirectors right now.

Reply:
👍